bluetooth: controller: Extend DTM commands

KAGA164 opened the pull request:
This PR adds a support for following HCI commands:

- HCI LE Receiver Test [v3]
- HCI LE Transmitter Test [v3]
- HCI LE Transmitter Test [v4]

Those command set add a possibility to test the radio CTE reception and transmission. The HCI LE Transmitter Test v4 commands allows setting a transmit power. The Controller under the DTM test can also generate the HCI Connectionless IQ Report event when it receives the  CTE.

Commit:
bluetooth: controller: Add support for all DTM commands

This adds support for the following Direct Test mode
commands:
- HCI LE Receiver Test [v3]
- HCI LE Transmitter Test [v3]
- HCI LE Transmitter Test [v4]

Those commands set add a possibility to test an CTE
reception and transmission. The HCI LE Transmitter
Test [v4] commands allows also setting a transmit power.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

Commit:
bluetooth: controller: Align the Openisa test ll API

Alignes the Openisa ll_test_tx and ll_test_rx functions
with the new test API.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

Commit:
bluetooth: samples: hci_uart: Add support for nrf52833dk

Adds support for the nrf52833dk_nrf52833 which supports
the Direction Finding feature. It allows also to perform
the Direct Test Mode with the CTE feature.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

cvinayak commented on samples/bluetooth/hci_uart/sample.yaml:
```diff
@@ -12,3 +12,16 @@ tests:
+    platform_allow: nrf52833dk_nrf52833
+    extra_configs:
+      - CONFIG_BT_CTLR_DF=y
+      - CONFIG_BT_CTLR_DTM_HCI_DF_IQ_REPORT=y"
```
```suggestion
      - CONFIG_BT_CTLR_DTM_HCI_DF_IQ_REPORT=y
```

KAGA164 replied:
Fixed

cvinayak commented on subsys/bluetooth/controller/Kconfig.dtm:
```diff
@@ -0,0 +1,49 @@
+	help
+	  Enable support for Direct Test Mode over the HCI transport.
+
+if BT_CTLR_DTM_HCI
+
+config BT_CTLR_DTM_HCI_RX_V3
```
```suggestion
if BT_CTLR_DTM_HCI

if BT_CTLR_DF_SUPPORT

config BT_CTLR_DTM_HCI_RX_V3
```

KAGA164 replied:
I am not sure about adding `BT_CTLR_DF_SUPPORT` dependency here. Those command can be still used in case if controller doesn't support DF. The spec has `Missing parameters` paragraph for those command that describes how to pass parameters when we for example don't want to use CTE. The transmitter command in version 4 is the only command which can set Tx power level, in my opinion it can be used even if Controller doesn't support DF and CTE length is set to 0, if the CTE length will be set to value other that 0 then it returns HCI ERROR CMD DISALLOWED. What do you thing about it ?

cvinayak replied:
fair enough.

cvinayak commented on subsys/bluetooth/controller/Kconfig.dtm:
```diff
@@ -0,0 +1,49 @@
+	select EXPERIMENTAL
+	help
+	  Enable generation of the HCI LE Connectionless IQ Report event
+	  after the Constant Tone Extension reception.
+
+endif # BT_CTLR_DTM_HCI
```
```suggestion
	  after the Constant Tone Extension reception.

endif #BT_CTLR_DF_SUPPORT

endif # BT_CTLR_DTM_HCI
```

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2795,15 +2807,18 @@ static void le_df_connectionless_iq_report(struct pdu_data *pdu_rx,
 	 */
-	if (ull_df_sync_cfg_is_not_enabled(&lll->df_cfg)) {
-		/* Dropp further processing of the event. */
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX)
+	if (lll && ull_df_sync_cfg_is_not_enabled(&lll->df_cfg)) {
+		/* Drop further processing of the event. */
```
Heads up, will conflict with #42302

KAGA164 replied:
Thanks. Resloved

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -85,6 +112,49 @@ static const uint8_t prbs15[255] = { 0x00, };
+#endif /* CONFIG_BT_CTLR_DF_CTE_RX */
+
+#if defined(CONFIG_BT_CTLR_DTM_HCI_DF_IQ_REPORT)
+static void create_iq_report(bool crc_ok)
```
```suggestion
static int create_iq_report(bool crc_ok)
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -85,6 +112,49 @@ static const uint8_t prbs15[255] = { 0x00, };
+
+	cte_info = radio_df_cte_status_get();
+	ant = radio_df_pdu_antenna_switch_pattern_get();
+	iq_report = ull_df_iq_report_alloc();
```
```suggestion
	iq_report = ull_df_iq_report_alloc();
	if (!iq_report) {
		return -ENOBUFS;
	}
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -85,6 +112,49 @@ static const uint8_t prbs15[255] = { 0x00, };
+	 * not detected and sampling was not started.
+	 */
+	if (!sample_cnt) {
+		return;
```
```suggestion
		return -EINVAL;
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -85,6 +112,49 @@ static const uint8_t prbs15[255] = { 0x00, };
+	ftr->param = NULL;
+	ftr->rssi = BT_HCI_LE_RSSI_NOT_AVAILABLE;
+
+	ull_rx_put(iq_report->hdr.link, iq_report);
+}
```
```suggestion
	ull_rx_put(iq_report->hdr.link, iq_report);
	
	return 0;
}
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+	}
+
+#if defined(CONFIG_BT_CTLR_DTM_HCI_DF_IQ_REPORT)
+	if (cte_ok) {
+		create_iq_report(crc_ok);
+		ull_rx_sched();
+	}
```
```suggestion
	if (cte_ok) {
		int err;

		err = create_iq_report(crc_ok);
		if (!err) {
			ull_rx_sched();
		}
	}
```

KAGA164 replied:
Good idea. Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+#endif
+}
+
+static uint32_t tx_power_check(int8_t tx_power)
```
```suggestion
static uint8_t tx_power_check(int8_t tx_power)
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+
+static uint32_t tx_power_set(int8_t tx_power)
+{
+	uint32_t err;
```
```suggestion
	uint8_t err;
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+	return 0;
+}
+
+static uint32_t tx_power_set(int8_t tx_power)
```
```suggestion
static uint8_t tx_power_set(int8_t tx_power)
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
 
-static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
+#if defined(CONFIG_BT_CTLR_DF_CTE_TX)
+static uint32_t cte_tx_parameters_check(uint8_t cte_len, uint8_t cte_type,
+					uint8_t switch_pattern_len)
```
```suggestion
static uint8_t cte_tx_parameters_check(uint8_t cte_len, uint8_t cte_type,
					uint8_t switch_pattern_len)
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+	return 0;
+}
+
+static uint32_t cte_tx_init(uint8_t cte_len, uint8_t cte_type, uint8_t switch_pattern_len,
+			    const uint8_t *ant_id)
```
```suggestion
static uint8_t cte_tx_init(uint8_t cte_len, uint8_t cte_type, uint8_t switch_pattern_len,
			    const uint8_t *ant_id)
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+static uint32_t cte_tx_init(uint8_t cte_len, uint8_t cte_type, uint8_t switch_pattern_len,
+			    const uint8_t *ant_id)
+{
+	uint32_t err;
```
```suggestion
	uint8_t err;
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+#endif /* CONFIG_BT_CTLR_DF_CTE_TX */
+
+#if defined(CONFIG_BT_CTLR_DF_CTE_RX)
+static uint32_t cte_rx_parameters_check(uint8_t expected_cte_len, uint8_t expected_cte_type,
+					uint8_t slot_duration, uint8_t switch_pattern_len)
```
```suggestion
static uint8_t cte_rx_parameters_check(uint8_t expected_cte_len, uint8_t expected_cte_type,
					uint8_t slot_duration, uint8_t switch_pattern_len)
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+	return 0;
+}
+
+static uint32_t cte_rx_init(uint8_t expected_cte_len, uint8_t expected_cte_type,
+			    uint8_t slot_duration, uint8_t switch_pattern_len,
+			    const uint8_t *ant_ids)
```
```suggestion
static uint8_t cte_rx_init(uint8_t expected_cte_len, uint8_t expected_cte_type,
			    uint8_t slot_duration, uint8_t switch_pattern_len,
			    const uint8_t *ant_ids)
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+			    uint8_t slot_duration, uint8_t switch_pattern_len,
+			    const uint8_t *ant_ids)
+{
+	uint32_t err;
```
```suggestion
	uint8_t err;
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+}
+#endif /* CONFIG_BT_CTLR_DF_CTE_RX */
+
+static uint32_t init(uint8_t chan, uint8_t phy, int8_t tx_power,
+		     bool cte, void (*isr)(void *))
```
```suggestion
static uint8_t init(uint8_t chan, uint8_t phy, int8_t tx_power,
		     bool cte, void (*isr)(void *))
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
 	radio_pkt_tx_set(pdu);
+}
+
+uint32_t ll_test_tx(uint8_t chan, uint8_t len, uint8_t type, uint8_t phy,
+		    uint8_t cte_len, uint8_t cte_type, uint8_t switch_pattern_len,
+		    const uint8_t *ant_id, int8_t tx_power)
```
```suggestion
uint8_t ll_test_tx(uint8_t chan, uint8_t len, uint8_t type, uint8_t phy,
		    uint8_t cte_len, uint8_t cte_type, uint8_t switch_pattern_len,
		    const uint8_t *ant_id, int8_t tx_power)
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
+		    uint8_t cte_len, uint8_t cte_type, uint8_t switch_pattern_len,
+		    const uint8_t *ant_id, int8_t tx_power)
+{
+	uint32_t start_us;
+	uint32_t err;
+	bool cte_request = (cte_len > 0) ? 1 : 0;
```
```suggestion
	const bool cte_request = (cte_len > 0) ? 1 : 0;
	uint32_t start_us;
	uint8_t err;
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
 	radio_tmr_tifs_set(150);
-	radio_tx_power_max_set();
+
+	err = tx_power_set(tx_power);
```
declare a local `uint8_t ret;`

```suggestion
	ret = tx_power_set(tx_power);
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
+			    RADIO_PKT_CONF_CTE(cte ? RADIO_PKT_CONF_CTE_ENABLED :
+						     RADIO_PKT_CONF_CTE_DISABLED));
+
+	return err;
```
```suggestion
	return ret;
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -281,19 +632,49 @@ uint32_t ll_test_tx(uint8_t chan, uint8_t len, uint8_t type, uint8_t phy)
 }
 
-uint32_t ll_test_rx(uint8_t chan, uint8_t phy, uint8_t mod_idx)
+uint32_t ll_test_rx(uint8_t chan, uint8_t phy, uint8_t mod_idx, uint8_t expected_cte_len,
+		    uint8_t expected_cte_type, uint8_t slot_duration, uint8_t switch_pattern_len,
+		    const uint8_t *ant_ids)
```
```suggestion
uint8_t ll_test_rx(uint8_t chan, uint8_t phy, uint8_t mod_idx, uint8_t expected_cte_len,
		    uint8_t expected_cte_type, uint8_t slot_duration, uint8_t switch_pattern_len,
		    const uint8_t *ant_ids)
```

KAGA164 replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -281,19 +632,49 @@ uint32_t ll_test_tx(uint8_t chan, uint8_t len, uint8_t type, uint8_t phy)
+		    uint8_t expected_cte_type, uint8_t slot_duration, uint8_t switch_pattern_len,
+		    const uint8_t *ant_ids)
 {
 	uint32_t err;
+	bool cte_expected = (expected_cte_len > 0) ? 1 : 0;
```
```suggestion
	const bool cte_expected = (expected_cte_len > 0) ? 1 : 0;
	uint8_t err;
```

KAGA164 replied:
Done

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
 	switch (type) {
 	case 0x00:
-		memcpy(payload, prbs9, len);
+		memcpy(pdu->payload, prbs9, len);
```
```suggestion
		(void)memcpy(pdu->payload, prbs9, len);
```
The same should be done for all `memset` occurrences below.

KAGA164 replied:
Ok added in all places

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
+{
+	uint32_t start_us;
+	uint32_t err;
+	bool cte_request = (cte_len > 0) ? 1 : 0;
```
```suggestion
	bool cte_request = (cte_len > 0U) ? 1U : 0U;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
+	uint32_t err;
+	bool cte_request = (cte_len > 0) ? 1 : 0;
+
+	if ((type > 0x07) || !phy || (phy > 0x04)) {
```
Change magic numbers to macros.

KAGA164 replied:
This isn't a part of this PR but I defined most of the magic number in this source file now

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
+	/* Configure Constant Tone Extension */
+	if (cte_request) {
+#if defined(CONFIG_BT_CTLR_DF_CTE_TX)
+		if (phy > BT_HCI_LE_PHY_2M) {
```
```suggestion
		if (phy == BT_HCI_LE_PHy_CODED) {
```

ppryga replied:
I didn't spot that.
In my humble opinion this may lead to some misunderstanding.

`phy` parameters for this command and other PHY related have common subset of values.
https://github.com/zephyrproject-rtos/zephyr/blob/d353e551607926612ff258dadbc13d028d58b54e/include/bluetooth/hci.h#L1189-L1191
But for PHY CODED there is a difference. Here PHY Coded means S2 and S8.
CODED S2 and S8 are defined separately:
https://github.com/zephyrproject-rtos/zephyr/blob/d353e551607926612ff258dadbc13d028d58b54e/include/bluetooth/hci.h#L1218-L1220.

I suggested to create a separate macros defining all allowed PHYs for the command and use it as it is now in the commented line of code.

KAGA164 replied:
There are also:
```
#define BT_HCI_LE_TX_PHY_CODED_S8               0x03
#define BT_HCI_LE_TX_PHY_CODED_S2               0x04
```

KAGA164 replied:
Added two macros sets for Tx and Rx test

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -281,19 +632,49 @@ uint32_t ll_test_tx(uint8_t chan, uint8_t len, uint8_t type, uint8_t phy)
 	uint32_t err;
+	bool cte_expected = (expected_cte_len > 0) ? 1 : 0;
 
 	if (!phy || (phy > 0x03)) {
```
I was not part of the changes in the PR but please change magic number to a macro.

KAGA164 replied:
Done

ppryga commented on samples/bluetooth/hci_uart/boards/nrf52833dk_nrf52833.overlay:
```diff
@@ -0,0 +1,33 @@
+/*
```
Wouldn't it be better to create e.g. nrf52833df_nrf52833_dtm.overlay and use it through `DTC_OVERLAY_FILE` CMAKE?
I'm not sure if DF overlay should be set to every build of `hci_uart` for nRF52833 target.

Also there should be similar configuration for nRF5340 board.

KAGA164 replied:
I don't have a strong opinion about that. It seems like we can keep radio antenna configuration safely in default board overlay since it doesn't have any impact on the Kconfig or source files, when building with DF feature disabled but maybe I missed something. I added also antenna configuration for `nrf5340dk_nrf5340_cpunet` target. I can prepare separated overlays files for this if you want.

ppryga replied:
In my humble opinion `hci_uart` should be a general sample that is oriented for common functionality, hence the DF and test commands should be enabled optionally (explicitly by end user).

KAGA164 replied:
Ok I added a special target like nrf5340dk_nrf5340_cpunet_df. You can build this sample with df support by invoking for example: `west build -b nrf5340dk_nrf5340_cpunet@df`

Commit:
bluetooth: samples: hci_uart: Add nrf5340 antenna description

Adds antenna description for the nrf5340dk_nrf5340_cpunet
target. It allows to performthe Direct Test Mode with the
CTE feature on this target.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

KAGA164 commented on include/bluetooth/hci.h:
```diff
@@ -2019,6 +2044,29 @@ struct bt_hci_rp_le_read_iso_link_quality {
+#define BT_HCI_TX_TEST_POWER_MIN_SET 0x7E
+#define BT_HCI_TX_TEST_POWER_MAX_SET 0x7F
+
+struct bt_hci_cp_le_tx_test_v4_tx_power {
```
Sure. Done

KAGA164 commented on subsys/bluetooth/controller/Kconfig.dtm:
```diff
@@ -0,0 +1,49 @@
+
+config BT_CTLR_DTM_HCI_DF_IQ_REPORT
+	bool "Connectionless IQ report HCI event"
+	depends on BT_CTLR_DF_CTE_RX
```
Added in way suggested here. I am not sure if commands in version 3 and 4 should depends on the DF feature

ppryga replied:
Do the tests commands require IQ reports if used in context without DF?

KAGA164 replied:
The IQ reports can be generated only on the CTE reception but even then IQ reports generation is not mandatory. In case when there is no CTE in PDUs IQ reports won't be generated even when IQ reports generation is enabled in Kconfig. If your device doesn't support CTE but you won't to play with different Tx power values you can use test tx command v4 but you need to send it with CTE length set to 0 which means that CTE is disabled, if you set any other valid CTE length for such device then you get HCI not supported feature error.

The Kconfig for it looks in following way now:
```
config BT_CTLR_DTM_HCI_DF_IQ_REPORT
	bool "Connectionless IQ report HCI event [EXPERIMENTAL]"
	depends on BT_CTLR_DF_CTE_RX && BT_CTLR_DTM_HCI_RX_V3
	select EXPERIMENTAL
	help
	  Enable generation of the HCI LE Connectionless IQ Report event
	  after the Constant Tone Extension reception.
````

KAGA164 commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+		return BT_HCI_ERR_UNSUPP_FEATURE_PARAM_VAL;
+	}
+
+	return 0;
```
Done

KAGA164 commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+		return BT_HCI_ERR_UNSUPP_FEATURE_PARAM_VAL;
+	}
+
+	return 0;
```
Done

KAGA164 commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+		return BT_HCI_ERR_UNSUPP_FEATURE_PARAM_VAL;
+	}
+
+	return 0;
```
Done

KAGA164 commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+
+	radio_df_iq_data_packet_set(node_rx->pdu, IQ_SAMPLE_TOTAL_CNT);
+#else
+	radio_df_iq_data_packet_set(NULL, 0);
```
Sending events with IQ samples is an optional controller feature according to the BLE spec 5.3. We should count packet as valid if CTE type and length is the same as expected in the Rx test

KAGA164 commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
-	radio_pkt_configure(RADIO_PKT_CONF_LENGTH_8BIT, 255, RADIO_PKT_CONF_PHY(test_phy));
-
-	return 0;
+	radio_pkt_configure(RADIO_PKT_CONF_LENGTH_8BIT, 255,
```
Good idea. Done

KAGA164 commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +499,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
-	pdu = radio_pkt_scratch_get();
-	payload = &pdu[2];
+#if defined(CONFIG_BT_CTLR_DF_CTE_TX)
+	pdu->cp = cte_len ? 1 : 0;
```
Done

KAGA164 commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -153,16 +242,227 @@ static void isr_rx(void *param)
+		if (test_cte_len > 0) {
+			cte_ok = check_rx_cte(cte_ready);
+			if (cte_ok) {
+				test_num_rx++;
```
According to BLE spec 5.3 we should discard packet which has CTE type and length different than expected  so we cannot count it as valid PDU

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3065,7 +3097,7 @@ static void le_tx_test(struct net_buf *buf, struct net_buf **evt)
 
 	status = ll_test_tx(cmd->tx_ch, cmd->test_data_len, cmd->pkt_payload,
-			    0x01);
+			    0x01, 0, 0, 0, NULL, BT_HCI_TX_TEST_POWER_MAX_SET);
```
We aim to do not use magic numbers everywhere it is possible. That eases reading and understanding a code. E.g. `cte_length` argument may use `BT_HCI_LE_NO_CTE_TX` instead of `0`.

KAGA164 replied:
I'm already using `BT_HCI_LE_TEST_CTE_DISABLED` for cte_length.

KAGA164 replied:
Done, all magic numbers are replaced by macros

ppryga-nordic reviewed (changes requested):
One new comment added. 
Please go through unresolved comments from previous review also.

KAGA164 commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2819,16 +2839,33 @@ static void le_df_connectionless_iq_report(struct pdu_data *pdu_rx,
+		sep->per_evt_counter = sys_cpu_to_le16(lll->event_counter);
+	}
+#endif /* CONFIG_BT_CTLR_DF_SCAN_CTE_RX */
+
```
Done. It requires two additional local variables but I also wanted to avoid conditional compilation occurrences

KAGA164 commented on subsys/bluetooth/shell/ll.c:
```diff
@@ -92,7 +92,7 @@ int cmd_test_rx(const struct shell *sh, size_t  argc, char *argv[])
 	mod_idx = strtoul(argv[3], NULL, 16);
 
-	err = ll_test_rx(chan, phy, mod_idx);
+	err = ll_test_rx(chan, phy, mod_idx, 0, 0, 0, 0, NULL);
```
Done

KAGA164 commented on subsys/bluetooth/shell/ll.c:
```diff
@@ -69,7 +69,7 @@ int cmd_test_tx(const struct shell *sh, size_t  argc, char *argv[])
 	phy  = strtoul(argv[4], NULL, 16);
 
-	err = ll_test_tx(chan, len, type, phy);
+	err = ll_test_tx(chan, len, type, phy, 0, 0, 0, NULL, 0x7F);
```
Done

KAGA164 commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3087,21 +3119,65 @@ static void le_enh_rx_test(struct net_buf *buf, struct net_buf **evt)
 
 	status = ll_test_tx(cmd->tx_ch, cmd->test_data_len, cmd->pkt_payload,
-			    cmd->phy);
+			    cmd->phy, 0, 0, 0, NULL, BT_HCI_TX_TEST_POWER_MAX_SET);
```
Done

KAGA164 commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3087,21 +3119,65 @@ static void le_enh_rx_test(struct net_buf *buf, struct net_buf **evt)
 	uint8_t status;
 
-	status = ll_test_rx(cmd->rx_ch, cmd->phy, cmd->mod_index);
+	status = ll_test_rx(cmd->rx_ch, cmd->phy, cmd->mod_index, 0, 0, 0, 0, NULL);
```
Done

KAGA164 commented:
@ppryga @cvinayak any comments ? Are we ready to merge it ?

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:
```diff
@@ -193,74 +517,119 @@ static uint32_t init(uint8_t chan, uint8_t phy, void (*isr)(void *))
+{
+	uint32_t start_us;
+	uint8_t err;
+	const bool cte_request = (cte_len > 0) ? 1U : 0U;
```
NIt-pick: 
```suggestion
	const bool cte_request = (cte_len > 0) ? true : false;
```
use `true`/`false` for `bool` variables.

KAGA164 replied:
Done in all places

Commit:
bluetooth: samples: hci_uart: Add Direction Finding doc

Add the Direction Finding documentation section into
the sample README.rst file.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

KAGA164 commented:
@ppryga I replaced magic number in shell related files and also I added a doc entry about the Direction Finding in hci_uart sample doc.

ppryga commented on samples/bluetooth/hci_uart/README.rst:
```diff
@@ -125,3 +125,19 @@ application. To enable debug over RTT the debug configuration file can be used.
 
 Then attach RTT as described here: :ref:`Using Segger J-Link <Using Segger J-Link>`
+
+Support for the Direction Finding
```
This is a very limited information about DF. I suggest to add here more comprehensive information or reference to direction finding samples. 
Or change this to information about DTM mode that allows to test DF functionality with V3/V4 version of HCI test commands.

KAGA164 replied:
Ok I added a references to the Direction Finding samples

ppryga commented:
Approved in regard of changes I've requested.

cvinayak reviewed (changes requested):
Since you would be testing DTM, this is a request to include related suggestion mentioned here: https://github.com/zephyrproject-rtos/zephyr/pull/42866#discussion_r808691029

KAGA164 commented:
I added DNM. I just want to check one more thing

KAGA164 commented:
@ppryga @cvinayak I needed to rebase due to the API changed for:
`void radio_df_cte_rx_2us_switching(bool cte_info_in_s1, uint8_t phy);` and
`void radio_df_cte_rx_4us_switching(bool cte_info_in_s1, uint8_t phy)`

ppryga-nordic reviewed (approved):
Quick look - ACK
